File: Chapter2(Service_and_Action)/my_action/src/witch_client.cpp
#include<ros/ros.h>
#include<iostream>
#include<actionlib/client/simple_action_client.h>
#include<my_action/witchAction.h>
#include<actionlib/client/terminal_state.h>

int main(int argc, char **argv)
{
	/* code */
	ros::init(argc, argv, "witch_client");
	if (argc != 3)
	{
		/* code */
		ROS_WARN("Usage: witch_client <goal> <time_to_preempt>");
		return -1;
	}
	actionlib::SimpleActionClient<my_action::witchAction> ac("witch_action", true);

	ROS_INFO("Waiting for action server ...");

	ac.waitForServer();

	my_action::witchGoal goal;

	goal.count = atoi(argv[1]);

	ROS_INFO("Sending goal %d and preempt time of %d", goal.count, atoi(argv[2]));

	ac.sendGoal(goal);

	bool status = ac.waitForResult(ros::Duration(atoi(argv[2])));

	ac.cancelGoal();

	if (status)
	{
		/* code */
		actionlib::SimpleClientGoalState state = ac.getState();

		ROS_INFO("Action finished %s", state.toString().c_str());

		ac.cancelGoal();	

	}
	else
	{
		ROS_INFO("Action did not finish before the time out");
	}
	return 0;
}

File: Chapter2(Service_and_Action)/my_action/scripts/witch_client.py
#! /usr/bin/env python

import rospy
import actionlib
from my_action.msg import witchAction, witchGoal

if __name__ == '__main__':
	rospy.init_node('witch_client')
	client = actionlib.SimpleActionClient('witch_client', witchAction)
	client.wait_for_server()

	goal = witchGoal()
	goal.count = 1000
	client.send_goal(goal)
	client.wair_for_result(rospy.Duration.from_sec(50.0))